### DISCUSSION TITLE
Create method not found on command

### DISCUSSION THREAD

**Discussion by ValentijnMakkenze** [MarimerLLC/csla#2456] [Questions]
**Describe the bug**
We get an exception after rewriting a command to get rid of the analyzer warning CSLA0011 CSLA business objects should not be created outside of a ObjectFactory instance.
We used constructors for our commands and not Create methods but after converting it the method is not called instead we get an "Invalid operation - create not allowed" exception. It looks like the objectarray is converted to an int and it can't find a create with int so it calls the base create which throws the exception.

**Version and Platform**
CSLA version: 5.5.1
OS: Windows
Platform: WinForms

**Code that Fails**

    [TestFixture]
    public class TestCommandLoadProjectDataBegin
    {
      [Test]
      public void TestCreate()
      {
        var command = Csla.DataPortal.Create<CommandLoadProjectData>(new[] { 3 });
        Assert.That(command, Is.Not.Null);
      }
    }

    [global::System.SerializableAttribute()]
    public class CommandLoadProjectDataBegin : Csla.CommandBase<CommandLoadProjectDataBegin>
    {
      private object[] PrimaryKey {get;set;}
    
      public string Data {get;set;}
    
      [Csla.CreateAttribute()]
      [Csla.RunLocalAttribute()]
      protected virtual void Create(object[] key)
      {
        this.PrimaryKey = key;
      }
    
      [Csla.ExecuteAttribute()]
      protected override void DataPortal_Execute()
      {
        // Not needed for repro
      }
    }

**Stack Trace or Exception Detail**
Csla.DataPortalException : DataPortal.Create failed (Invalid operation - create not allowed)
  ----> Csla.Reflection.CallMethodException : CommandLoadProjectData.DataPortal_Create method call failed
  ----> System.NotSupportedException : Invalid operation - create not allowed
   at Csla.CommandBase`1.DataPortal_Create(Object criteria)
   at lambda_method(Closure , Object , Object[] )
   at Csla.Reflection.ServiceProviderMethodCaller.<CallMethodTryAsync>d__11.MoveNext()
   at Csla.Reflection.ServiceProviderMethodCaller.<CallMethodTryAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Reflection.LateBoundObject.<CallMethodTryAsyncDI>d__12`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.DataPortalTarget.<InvokeOperationAsync>d__14`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.DataPortalTarget.<CreateAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.SimpleDataPortal.<Create>d__0.MoveNext()
   at Csla.DataPortal`1.Create(Type objectType, Object criteria)
   at Csla.DataPortal`1.Create(Object[] criteria)

**Comment by ValentijnMakkenze** [comment]
We have a similar problem with Create methods that expect a long but get an int.

So this fails

      var command = DataPortal.Create<HeeftAppartementenCommand>(17, 22222);
      
But this does not fail (added L for explicit long)

      var command = DataPortal.Create<HeeftAppartementenCommand>(17L, 22222L);

**Comment by rockfordlhotka** [comment]
Right now the [FindDataPortalMethod](https://github.com/MarimerLLC/csla/blob/613b8795a85aa0ce32ae96b6e975c74d506df0f6/Source/Csla/Reflection/ServiceProviderMethodCaller.cs#L54) method looks for exact parameter matches, falling back to type `object`. It doesn't currently fall back to, or test for, assignability of one type to another.

I suppose maybe it could do that, assigning a lower score to an assignable type compared to a direct type match.

In both cases @ValentijnMakkenze, what is extremely useful is to have a unit test in the [ServiceProviderMethodCallerTests](https://github.com/MarimerLLC/csla/blob/v5.x/Source/csla.netcore.test/DataPortal/ServiceProviderMethodCallerTests.cs) class that demonstrates the failure and desired behavior.

**Comment by rockfordlhotka** [comment]
@ValentijnMakkenze I'm looking at your code

```c#
protected virtual void Create(object[] key)
```

I don't think this is, or can be, supported.

The client-side `CreateAsync` method accepts a params array, and when you pass in an array like in your example:

```c#
var command = Csla.DataPortal.Create<CommandLoadProjectData>(new[] { 3 });
```

C# passes that in as an array of params with one element. I don't think it can be distinguished from `Create<>(3);` - which also comes in as a single element array. 

Specifically, your `object[]` parameter is a _direct match_ for the params array in CSLA:

```c#
    public T Create(params object[] criteria)
```

C# will clearly take your `object[]` parameter and map it to that method overload.

So _in both cases_ the server-side method signature will be resolved to:

```c#
  [Create]
  private void Create(int x)
```

If you really want to pass an arbitrary list of untyped values as a parameter to a data portal method, you should use a `MobileList`:

```c#
var command = DataPortal.Create<CommandLoadProjectData>(new Csla.Core.MobileList { 3 });
```

and

```c#
  [Create]
  private void Create(Csla.Core.MobileList criteria)
```
